src/Symfony/Component/Lock/BlockingStoreInterface.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock;
+
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+
+/**
+ * @author Hamza Amrouche <hamza.simperfit@gmail.com>
+ */
+interface BlockingStoreInterface
+{
+    /**
+     * Waits until a key becomes free, then stores the resource.
+     *
+     * If the store does not support this feature it should throw a NotSupportedException.
+     *
+     * @throws LockConflictedException
+     * @throws NotSupportedException
+     */
+    public function waitAndSave(Key $key);
+
+    /**
+     * Checks if the store can wait until a key becomes free before storing the resource.
+     */
+    public function supportsWaitAndSave(): bool;
```

Comment:
I don't understand the purpose of this method. 
`waitAndSave()` is the only method of the interface, implementing the interface without supporting it does not make sense. 
An `instanceof BlockingStoreInterface` check should be enough to call `waitAndSave()` safely.

Reply:
It can safely return true, or return false in some case, using only an instance of BlockingStore seems to not allow people to decide if it supports it by making some tests.

Reply:
@chalasr it make sens for decorator, where the decorated service may not implement the interface.

Reply:
Not convinced, really. A decorator should either not implement the optional interface or have a behavior that varies depending on if the decorated implements the interface or not (i.e. if not, make it a no-op, otherwise forward the call to the decorated).
See e.g Cache's `TagAwareAdapterInterface` which is similar and doesn't have such method. Returning false here is contradictory, it defeats the purpose of having a dedicated interface IMHO. /cc @symfony/deciders

Reply:
Sadly it don't scale.. See `TagAwareAdapterInterface` which imply `class TraceableTagAwareAdapter extends TraceableAdapter`. With 2 sub-interfaces (like it was in the first commit of that PR) we would have to implement 4 decorators `Foo`, `FooWithA`,  `FooWithB`, `FooWithAAndB`...  

That being said, now that `ExpiringInterface` is gone, the current component have 2 decorators:
- `RetryTillSaveStore` which bring the `waitAndSave` behavior.
- `CombinedStore` that cannot support `waitAndSave` (even if decorated store does)

Meaning that *we* don't need this method which can be removed: all *our* implementations and decorators are either fully compatible with the interface or incompatible with the interface:  we don't have case where the compatibility depends on something else.

Reply:
updated in #32492